src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -24605,6 +24605,1499 @@ void Init()
+            _ => throw null!,
+        };
+    }
+    
+    string M2(string s) => throw null!;
+}
+", MemberNotNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        [WorkItem(51020, "https://github.com/dotnet/roslyn/issues/51020")]
+        public void MemberNotNullWhenTrue_TupleInIsExpression()
+        {
+            var c = CreateCompilation(new[] { @"
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public class Test
+{
+    [MemberNotNullWhen(true, nameof(Value))]
+    public bool IsOk { get; }
+    public string? Value { get; }
+
+    public string M(object? o)
+    {
+        if ((this, o) is ({ IsOk: true }, null))
+        {
+            return Value;
+        }
+
+        throw null!;
+    }
+}
+", MemberNotNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        [WorkItem(50980, "https://github.com/dotnet/roslyn/issues/50980")]
+        public void MemberNotNullWhenTrue_TupleEquality()
+        {
+            var c = CreateCompilation(new[] { @"
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public class Test
+{
+    [MemberNotNullWhen(true, nameof(Value))]
+    public bool IsOk { get; }
+    public string? Value { get; }
+
+    public string M(object? o)
+    {
+        if ((IsOk, o) is (true, null))
+        {
+            return Value; // incorrect warning
+        }
+
+        if (IsOk is true)
+        {
+            return Value;
+        }
+
+        throw null!;
+    }
+}
+", MemberNotNullWhenAttributeDefinition });
+
+            // We're not learning from tuple equality...
+            // Tracked by https://github.com/dotnet/roslyn/issues/50980
+            c.VerifyDiagnostics(
+                // (15,20): warning CS8603: Possible null reference return.
+                //             return Value; // incorrect warning
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "Value").WithLocation(15, 20)
+                );
+        }
+
+        [Fact, WorkItem(51020, "https://github.com/dotnet/roslyn/issues/51020")]
+        public void PatternOnTuple_IsExpression()
+        {
+            var src = @"
+#nullable enable
+public class C
+{
+    void M(object? o, object o2)
+    {
+        if ((o, o2) is (not null, null))
+        {
+            o.ToString();
+            o2.ToString(); // 1
+        }
+    }
+
+    void M2(object? o)
+    {
+        if (o is not null)
+        {
+            o.ToString();
+        }
+    }
+
+    void M3(object o2)
+    {
+        if (o2 is null)
+        {
+            o2.ToString(); // 2
+        }
+    }
+
+    void M4(object? o, object o2)
+    {
+        if ((true, (o, o2)) is (true, (not null, null)))
+        {
+            o.ToString(); // 3
+            o2.ToString();
```

Comment:
Shouldn't the warning be reported for `o2` rather than `o`?

Reply:
We're dealing with a nested tuple here, which makes us lose track. So the test `o is not null` does not update the state of `o` (that's why we still get a warning on `o.ToString()`). Similarly, the pure null test on `o2` doesn't actually affect `o2`, but instead some temp.
I'll add a comment to clarify that the behavior is not what we'd prefer.
Let me know if you think I should file a follow-up issue. I didn't prioritize this scenario because I think nested tuples are not very useful in pattern-matching...

---
In reply to: [573098133](https://github.com/dotnet/roslyn/pull/51001#discussion_r573098133) [](ancestors = 573098133)